File: benchmark/go-bench/internal/clientbench/clientbench.go
// Package clientbench holds the shared driving logic for the Go client
// benchmarks (net/http, fasthttp): CLI parsing, the concurrent load loop,
// HdrHistogram latency percentiles, and the harness result line. Each concrete
// client supplies only a `get(url) error` that performs one request and
// consumes its body.
//
// Latency percentiles use HdrHistogram with 3 significant figures — the same
// method the JVM (org.HdrHistogram) and Rust (hdrhistogram crate) drivers use,
// so all drivers report directly comparable p50/p99/p99.9. The number of
// significant figures determines the bucketing, so any HdrHistogram port with
// sigfigs=3 yields identical percentiles regardless of its tracked value range.
package clientbench

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/HdrHistogram/hdrhistogram-go"
)

// Latency histogram bounds shared by every driver: 1 ns to 60 s, 3 significant
// figures. The range only bounds allocation (no realistic request latency
// exceeds it); sigfigs=3 is what fixes the percentile values across drivers.
const (
	latLowest  = 1
	latHighest = 60_000_000_000
	latSigFigs = 3
)

func newLatencyHist() *hdrhistogram.Histogram {
	return hdrhistogram.New(latLowest, latHighest, latSigFigs)
}

// Config mirrors the JVM harness client flags.
type Config struct {
	Targets     []string // full URLs (base + endpoint)
	Endpoint    string
	Connections int
	Warmup      int
	Duration    int
	Pinned      bool
}

// Parse reads the --client-* flags shared with the JVM harness.
func Parse() Config {
	target := ""
	endpoint := "/hello"
	connections := 50
	warmup := 3
	duration := 10
	pinned := false
	for _, arg := range os.Args[1:] {
		if !strings.HasPrefix(arg, "--") || !strings.Contains(arg, "=") {
			continue
		}
		kv := strings.SplitN(strings.TrimPrefix(arg, "--"), "=", 2)
		switch kv[0] {
		case "client-target":
			target = kv[1]
		case "client-endpoint":
			endpoint = kv[1]
		case "client-connections":
			connections, _ = strconv.Atoi(kv[1])
		case "client-warmup":
			warmup, _ = strconv.Atoi(kv[1])
		case "client-duration":
			duration, _ = strconv.Atoi(kv[1])
		case "client-target-mode":
			pinned = kv[1] == "pinned"
		}
	}
	if target == "" {
		fmt.Fprintln(os.Stderr, "missing --client-target")
		os.Exit(1)
	}
	var targets []string
	for _, t := range strings.Split(target, ",") {
		t = strings.TrimSpace(t)
		if t == "" {
			continue
		}
		targets = append(targets, strings.TrimRight(t, "/")+endpoint)
	}
	if len(targets) == 0 {
		fmt.Fprintln(os.Stderr, "no target parsed from --client-target")
		os.Exit(1)
	}
	return Config{targets, endpoint, connections, warmup, duration, pinned}
}

// Run drives get across Connections goroutines for secs seconds and returns the
// throughput, an HdrHistogram of the per-request latencies (ns), and the error
// count.
func Run(cfg Config, secs int, get func(url string) error) (float64, *hdrhistogram.Histogram, int64) {
	deadline := time.Now().Add(time.Duration(secs) * time.Second)
	var completed, errors int64
	var pick uint64
	perWorker := make([]*hdrhistogram.Histogram, cfg.Connections)

	var wg sync.WaitGroup
	start := time.Now()
	for w := 0; w < cfg.Connections; w++ {
		wg.Add(1)
		go func(worker int) {
			defer wg.Done()
			local := newLatencyHist()
			var pinnedURL string
			if cfg.Pinned {
				pinnedURL = cfg.Targets[worker%len(cfg.Targets)]
			}
			for time.Now().Before(deadline) {
				url := pinnedURL
				if url == "" {
					i := int(atomic.AddUint64(&pick, 1)-1) % len(cfg.Targets)
					url = cfg.Targets[i]
				}
				t0 := time.Now()
				if err := get(url); err != nil {
					atomic.AddInt64(&errors, 1)
					continue
				}
				_ = local.RecordValue(time.Since(t0).Nanoseconds())
				atomic.AddInt64(&completed, 1)
			}
			perWorker[worker] = local
		}(w)
	}
	wg.Wait()
	elapsed := time.Since(start).Seconds()

	total := newLatencyHist()
	for _, l := range perWorker {
		if l != nil {
			total.Merge(l)
		}
	}
	rps := 0.0
	if elapsed > 0 {
		rps = float64(completed) / elapsed
	}
	return rps, total, errors
}

func pctMs(h *hdrhistogram.Histogram, q float64) float64 {
	return float64(h.ValueAtQuantile(q)) / 1e6
}

// Report prints the harness result line (bytes/op is n/a for native clients).
func Report(name string, cfg Config, rps float64, lat *hdrhistogram.Histogram, errors int64) {
	maxMs := float64(lat.Max()) / 1e6
	fmt.Printf("%s%s|%.0f|%.3f|%.3f|%.3f|%.3f|n/a|%d\n",
		name, cfg.Endpoint, rps,
		pctMs(lat, 50), pctMs(lat, 99), pctMs(lat, 99.9), maxMs, errors)
}

// LogStart writes the stderr banner shared by the Go client benches.
func LogStart(name string, cfg Config) {
	fmt.Fprintf(os.Stderr, "client bench: type=%s targets=%d conns=%d warmup=%ds duration=%ds\n",
		name, len(cfg.Targets), cfg.Connections, cfg.Warmup, cfg.Duration)
}

File: benchmark/rust-bench/src/clientbench.rs
// Shared driving logic for the Rust client benchmarks (reqwest, hyper): CLI
// parsing, the concurrent load loop, and the harness result line. Each client
// implements [HttpGet] (one GET that consumes its body); everything else is
// shared. Included into each bin via `#[path = "../clientbench.rs"] mod ...`.

use hdrhistogram::Histogram;
use std::future::Future;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::time::{Duration, Instant};

pub struct Config {
    pub targets: Vec<String>, // full URLs (base + endpoint)
    pub endpoint: String,
    pub connections: usize,
    pub warmup: u64,
    pub duration: u64,
    pub pinned: bool,
}

pub fn parse() -> Config {
    let mut target = String::new();
    let mut endpoint = "/hello".to_string();
    let mut connections = 50usize;
    let mut warmup = 3u64;
    let mut duration = 10u64;
    let mut pinned = false;
    for arg in std::env::args().skip(1) {
        if let Some((k, v)) = arg.strip_prefix("--").and_then(|s| s.split_once('=')) {
            match k {
                "client-target" => target = v.to_string(),
                "client-endpoint" => endpoint = v.to_string(),
                "client-connections" => connections = v.parse().expect("bad connections"),
                "client-warmup" => warmup = v.parse().expect("bad warmup"),
                "client-duration" => duration = v.parse().expect("bad duration"),
                "client-target-mode" => pinned = v == "pinned",
                _ => {}
            }
        }
    }
    assert!(!target.is_empty(), "missing --client-target");
    let targets = target
        .split(',')
        .map(|t| t.trim())
        .filter(|t| !t.is_empty())
        .map(|t| format!("{}{}", t.trim_end_matches('/'), endpoint))
        .collect::<Vec<_>>();
    assert!(!targets.is_empty(), "no target parsed");
    Config { targets, endpoint, connections, warmup, duration, pinned }
}

/// A client that performs one GET and fully consumes the body, returning
/// whether it succeeded. Cloneable so every worker task shares the pool.
pub trait HttpGet: Clone + Send + Sync + 'static {
    fn get(&self, url: String) -> impl Future<Output = bool> + Send;
}

pub struct Phase {
    pub rps: f64,
    pub hist: Histogram<u64>,
    pub errors: u64,
}

pub async fn run_phase<C: HttpGet>(client: C, cfg: &Config, secs: u64) -> Phase {
    let deadline = Instant::now() + Duration::from_secs(secs);
    let completed = Arc::new(AtomicU64::new(0));
    let errors = Arc::new(AtomicU64::new(0));
    let pick = Arc::new(AtomicU64::new(0));
    let targets = Arc::new(cfg.targets.clone());

    let start = Instant::now();
    let mut handles = Vec::with_capacity(cfg.connections);
    for worker in 0..cfg.connections {
        let client = client.clone();
        let targets = targets.clone();
        let completed = completed.clone();
        let errors = errors.clone();
        let pick = pick.clone();
        let pinned = if cfg.pinned { Some(targets[worker % targets.len()].clone()) } else { None };
        handles.push(tokio::spawn(async move {
            let mut hist = Histogram::<u64>::new(3).unwrap();
            while Instant::now() < deadline {
                let url = match &pinned {
                    Some(u) => u.clone(),
                    None => {
                        let i = pick.fetch_add(1, Ordering::Relaxed) as usize % targets.len();
                        targets[i].clone()
                    }
                };
                let t0 = Instant::now();
                if client.get(url).await {
                    let _ = hist.record((t0.elapsed().as_nanos() as u64).max(1));
                    completed.fetch_add(1, Ordering::Relaxed);
                } else {
                    errors.fetch_add(1, Ordering::Relaxed);
                }
            }
            hist
        }));
    }

    let mut total = Histogram::<u64>::new(3).unwrap();
    for h in handles {
        if let Ok(hist) = h.await {
            let _ = total.add(&hist);
        }
    }
    let elapsed = start.elapsed().as_secs_f64().max(1e-9);
    Phase {
        rps: completed.load(Ordering::Relaxed) as f64 / elapsed,
        hist: total,
        errors: errors.load(Ordering::Relaxed),
    }
}

pub fn report(name: &str, cfg: &Config, p: &Phase) {
    let ms = |q: f64| p.hist.value_at_quantile(q) as f64 / 1e6;
    println!(
        "{}{}|{:.0}|{:.3}|{:.3}|{:.3}|{:.3}|n/a|{}",
        name,
        cfg.endpoint,
        p.rps,
        ms(0.50),
        ms(0.99),
        ms(0.999),
        p.hist.max() as f64 / 1e6,
        p.errors
    );
}

pub fn log_start(name: &str, cfg: &Config) {
    eprintln!(
        "client bench: type={} targets={} conns={} warmup={}s duration={}s",
        name,
        cfg.targets.len(),
        cfg.connections,
        cfg.warmup,
        cfg.duration
    );
}
